Documentation revisions for v3 (#3150)
geopm/geopm: service/docs/source/runtime.rst

support a wide range of tuning algorithms while solving three related
challenges:

1. Measuring Performance
^^^^^^^^^^^^^^^^^^^^^^^^

For hardware tuning algorithms, the first challenge involves generating a
durable estimate of application performance. In energy efficiency terms,

> performance measurements are power-to-performance ratios, for example,

Reviewer: ```suggestion
performance measurements are performance-to-power ratios, for example,
```

support a wide range of tuning algorithms while solving three related 
challenges:

1. Measuring Performance
^^^^^^^^^^^^^^^^^^^^^^^^

> For hardware tuning algorithms, the first challenge involves generating a durable estimate of application performance. In energy efficiency terms, performance measurements are power-to-performance ratios, for example, "perf per watt". Therefore, any dynamic hardware power control tuning that aims for energy efficiency must formulate an estimate of application performance. Without specific application feedback on the critical path, these performance estimates might prove inaccurate, causing hardware tuning algorithms to potentially disrupt application performance and lead to elongated run times, resulting in higher energy costs per unit of work than without the adaptive algorithm.

Reviewer 2: Column limit (more instances below)

2. Hardware Configuration
^^^^^^^^^^^^^^^^^^^^^^^^^

The second challenge arises from allowing hardware control algorithms to be influenced by unprivileged user input (like application feedback), which presents a security risk. Not having adequate checks can result in escalated privileges, service denial, and impacts on other system users' quality of service. GEOPM Service is created specifically to mitigate these issues.

3. Advanced Data Analysis
^^^^^^^^^^^^^^^^^^^^^^^^^

> The third challenge is to provide a software development platform suitable for control algorithms that can be securely deployed and use high-level languages for data analysis. To effectively enhance energy efficiency for certain applications, substantial software dependencies may be required. Control algorithms might lean on optimization software like machine learning packages or other numerical packages. The application being optimized could include millions of lines of code, and there may be significant interlinking between the application and the control algorithm. Limiting the privileges of process running the control algorithm significantly reduces software security audit requirements.

Reviewer 2: ```suggestion
The third challenge is to provide a software development platform suitable for control algorithms that can be securely deployed and use high-level languages for data analysis. To effectively enhance energy efficiency for certain applications, substantial software dependencies may be required. Control algorithms might lean on optimization software like machine learning or other numerical packages. The application being optimized could include millions of lines of code, and there may be significant interlinking between the application and the control algorithm. Limiting the privileges of the process running the control algorithm significantly reduces software security audit requirements.
```

Introduction
------------

GEOPM Runtime creates a bridge between GEOPM’s :doc:`application instrumentation interfaces <geopm_prof.3>` and :doc:`platform monitoring/control interfaces <geopm_pio.7>`.

By default, the GEOPM Runtime presents relationships between application instrumentation and platform-monitoring interfaces in a report. For more complex interactions, such as dynamic control of platform settings, different GEOPM *agents* can be utilized.  For more information on user-facing GEOPM Runtime launch options, please refer to :doc:`geopmlaunch(1)<geopmlaunch.1>` documentation.

.. figure:: https://geopm.github.io/images/geopm-runtime-usage.svg

>    :alt: An illustration of geopmlaunch running on 2 servers, generating a trace file per host, and one report across all hosts.

Reviewer 2: ```suggestion
   :alt: An illustration of ``geopmlaunch`` running on 2 servers, generating a trace file per host, and one report across all hosts.
```

   :align: center

.. admonition:: Quick start for MPI applications

>    The ``geopmlaunch`` tool is the recommended user interface for GEOPM Runtime. It wraps a launcher application
>    (like srun in this example), generates a summarizing report file, and optionally generates a time-series trace for each host.

Reviewer 2: ```suggestion
   The ``geopmlaunch`` tool is the recommended user interface for the GEOPM Runtime. It wraps a launcher application
   (like ``srun`` in this example), generates a summarizing report file, and optionally generates a time-series trace for each host.
```

GEOPM *agents* can exploit this hierarchical control system to optimize various objective functions. Examples include maximizing application performance within a power limit (such as GEOPM :doc:`power_balancer agent<geopm_agent_power_balancer.7>`) or decreasing energy consumption while minimally affecting application performance. The control hierarchy root can communicate with the system resource manager to extend the hierarchy beyond the individual MPI application, thus facilitating multiple MPI jobs and multiple-user system resource management.

The GEOPM Runtime package includes the libgeopm shared object library. GEOPM comes with numerous command-line tools, each with dedicated manual pages. The :doc:`geopmlaunch(1) <geopmlaunch.1>` command-line tool launches an MPI application, enabling the GEOPM runtime to create a GEOPM Controller thread on each compute node. The Controller loads plugins and runs the Agent algorithm to manage the compute application. The :doc:`geopmlaunch(1)<geopmlaunch.1>` command features in the geopmpy python package that is part of the GEOPM installation. For more documentation and links, please visit the :doc:`GEOPM overview man page <geopm.7>`.

> GEOPM Runtime offers several built-in algorithms, each incorporated within an "Agent" implementing the :doc:`geopm::Agent(3) <GEOPM_CXX_MAN_Agent.3>` class interface. Developers can expand these algorithm features by creating an Agent plugin. An implementation of this class can be dynamically loaded at runtime by the GEOPM Controller. The Agent class determines what data is collected, how control decisions are made, and how messages are exchanged between Agents in the compute nodes' tree hierarchy. The GEOPM Service package, which resides in the service directory of the GEOPM repository, provides the PlatformIO interface which abstracts data reading and writing controls from the Agent within a compute node. This allows Agent implementations to be ported to various hardware platforms without modification.

Reviewer 2: ```suggestion
GEOPM Runtime offers several built-in algorithms, each incorporated within an "Agent" implementing the :doc:`geopm::Agent(3) <GEOPM_CXX_MAN_Agent.3>` class interface. Developers can expand these algorithm features by creating an Agent plugin. An implementation of this class can be dynamically loaded at runtime by the GEOPM Controller. The Agent class determines what data is collected, how control decisions are made, and how messages are exchanged between Agents in the compute nodes' tree hierarchy. The GEOPM Service package, which resides in the service directory of the GEOPM repository, provides the PlatformIO interface which abstracts reading signals and writing controls from the Agent within a compute node. This allows Agent implementations to be ported to various hardware platforms without modification.
```

The libgeopm library can be called indirectly or directly within MPI applications, enabling application feedback to aid control decisions. Indirect calls are facilitated through GEOPM's integration with MPI and OpenMP via their profiling decorators. Direct calls are made through :doc:`geopm_prof(3)<geopm_prof.3>` or :doc:`geopm_fortran(3)<geopm_fortran.3>` interfaces. The application can be better integrated with the GEOPM runtime and controlled more accurately by marking up the compute application with profiling information obtained through these interfaces.

Build Requirements
------------------

> When building the GEOPM Runtime using the base build system in the GEOPM source repository, additional requirements must be met. Those uninterested in building the GEOPM Runtime can ignore these requirements, or by providing the disable flag to the base build configure command line, users may skip particular GEOPM Runtime features enabled by these requirements.

Reviewer 2: Minor recommendation:

```suggestion
When building the GEOPM Runtime from source, additional requirements must be met. Those uninterested in building the GEOPM Runtime can ignore these requirements, or by providing the disable flag to the configure command line, users may skip particular GEOPM Runtime features enabled by these requirements.
```

In general, we're moving away from using the term "base" build/directory in favor of the term runtime.

> The GEOPM Runtime requires MPI standards, Message Passing Interface, version 2.2 or later. Fulfilling this requirement generally depends on the specific HPC resource targeted based on site-specific documentation. The Intel MPI implementation, OpenHPC or Spack packaging systems, or OpenMPI binaries distributed with most major Linux distributions can satisfy this requirement. For RHEL and SLES Linux, the requirement can be met by installing the ``openmpi-devel`` package version 1.7 or later, and ``libopenmpi-dev`` on Ubuntu.

Reviewer 2: ```suggestion
The GEOPM Runtime requires MPI standards, Message Passing Interface, version 2.2 or later. Fulfilling this requirement generally depends on the specific HPC resource targeted based on site-specific documentation. The Intel MPI implementation, OpenHPC or Spack packaging systems, or OpenMPI binaries distributed with most major Linux distributions satisfy this requirement. For RHEL and SLES Linux, the requirement can be met by installing the ``openmpi-devel`` package version 1.7 or later, and ``libopenmpi-dev`` on Ubuntu.
```

:doc:`geopm_prof(3)<geopm_prof.3>` or :doc:`geopm_fortran(3)<geopm_fortran.3>`
interfaces. The application can be better integrated with the GEOPM runtime
and controlled more accurately by marking up the compute application with
profiling information obtained through these interfaces.

Build Requirements
------------------

When building the GEOPM Runtime from source, additional requirements must
be met. Those uninterested in building the GEOPM Runtime can ignore these
requirements, or by providing the disable flag to the configure command
line, users may skip particular GEOPM Runtime features enabled by these
requirements.

The GEOPM Runtime requires MPI standards, Message Passing Interface,
version 2.2 or later. Fulfilling this requirement generally depends on the
specific HPC resource targeted based on site-specific documentation. The
Intel MPI implementation, OpenHPC or Spack packaging systems, or OpenMPI
binaries distributed with most major Linux distributions satisfy this
requirement. For RHEL and SLES Linux, the requirement can be met by installing
the ``openmpi-devel`` package version 1.7 or later, and ``libopenmpi-dev``

> on Ubuntu.

Reviewer: "requires" -> "provides optional support for"

in general should reword this section.

Run Requirements
----------------

Beyond the GEOPM Service, the GEOPM Runtime requires several additional features at the time of use. Users uninterested in running the GEOPM Runtime can ignore these requirements.

.. contents:: Categories of run requirements:
   :local:

BIOS Configuration
^^^^^^^^^^^^^^^^^^

If power governing or power balancing is the intended usage for GEOPM deployment, an additional requirement involves configuring the BIOS to support RAPL control. To make this check for BIOS support, execute the following on a compute node:

.. code-block:: bash

    ./tutorial/admin/00_test_prereqs.sh

If the script output includes:

.. code-block:: none

    WARNING: The lock bit for the PKG_POWER_LIMIT MSR is set.  The power_balancer
             and power_governor agents will not function properly until this is cleared.

Please enable RAPL in your BIOS, and if such an option doesn't exist please contact your BIOS vendor to obtain a BIOS that supports RAPL.

For additional information, please contact the GEOPM team.

Linux Power Management
^^^^^^^^^^^^^^^^^^^^^^

> It's crucial to note that other Linux mechanisms for power management can interfere with GEOPM, which must be disenabled. Recommending disabling the intel_pstate kernel driver by modifying the kernel command line through grub2 or your system bootloader by adding:

Reviewer 2: ```suggestion
It's crucial to note that other Linux mechanisms for power management can interfere with GEOPM, which must be disabled. Recommending disabling the ``intel_pstate`` kernel driver by modifying the kernel command line through grub2 or your system bootloader by adding:
```

Also, check grammar in second sentence.

.. code-block:: bash

   "intel_pstate=disable"

> `Cpufreq` driver will be enabled when the intel_pstate driver is disabled. The cpufreq driver has several modes controlled by the scaling_governor sysfs entry. When the performance mode is selected, the driver will not interfere with GEOPM. On SLURM-based systems, the :ref:`GEOPM launch wrapper <runtime:geopm application launch wrapper>` will attempt to set the scaling governor to "performance" automatically, eliminating the need to manually set the governor. On older versions of SLURM, the desired governors must be listed explicitly in ``/etc/slurm.conf``. Specifically, SLURM 15.x requires the following option:

Reviewer 2: ```suggestion
The ``cpufreq`` driver will be enabled when the ``intel_pstate`` driver is disabled. It has several modes controlled by the ``scaling_governor`` sysfs entry. When the performance mode is selected, the driver will not interfere with GEOPM. On SLURM-based systems, the :ref:`GEOPM launch wrapper <runtime:geopm application launch wrapper>` will attempt to set the scaling governor to "performance" automatically, eliminating the need to manually set the governor. On older versions of SLURM, the desired governors must be listed explicitly in ``/etc/slurm.conf``. Specifically, SLURM 15.x requires the following option:
```

.. code-block:: bash

   CpuFreqGovernors=OnDemand,Performance

For more on SLURM configuration, please see the `slurm.conf manual <https://slurm.schedmd.com/slurm.conf.html>`_. On non-SLURM systems, the scaling governor should still be manually set through some other mechanism to ensure proper GEOPM behavior. The following command will set the governor to performance:

.. code-block:: bash

   echo performance | tee /sys/devices/system/cpu/cpu*/cpufreq/scaling_governor

For more information, see the Linux Kernel documentation on `cpu-freq governors <https://www.kernel.org/doc/Documentation/cpu-freq/governors.txt>`_.

GEOPM Application Launch Wrapper
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> The GEOPM Runtime package installs the ``geopmlaunch`` command. This command is a wrapper for MPI launch commands such as ``srun``, ``aprun``, and ``mpiexec``, where the wrapper script enables the GEOPM runtime. The geopmlaunch command supports the same command-line interface as the underlying launch command, while extending the interface with GEOPM-specific options. The ``geopmlaunch`` application launches the primary compute application and the GEOPM control thread on each compute node, and manages all process CPU affinity requirements. This wrapper is documented in the :doc:`geopmlaunch(1)<geopmlaunch.1>` man page.

Reviewer 2: ```suggestion
The GEOPM Runtime package installs the ``geopmlaunch`` command. This command is a wrapper for MPI launch commands such as ``srun``, ``aprun``, and ``mpiexec``, where the wrapper script enables the GEOPM runtime. The ``geopmlaunch`` command supports the same command-line interface as the underlying launch command, while extending the interface with GEOPM-specific options. The ``geopmlaunch`` application launches the primary compute application and the GEOPM control thread on each compute node, and manages all process CPU affinity requirements. This wrapper is documented in the :doc:`geopmlaunch(1)<geopmlaunch.1>` man page.
```

``Geopmlaunch`` supports various underlying MPI application launchers as shown in the :doc:`geopmlaunch(1)<geopmlaunch.1>` man page. If your system's launch mechanism is not supported, then you must enforce affinity requirements, and all options to the GEOPM runtime must be passed through environment variables. Please consult the :doc:`geopm(7)<geopm.7>` man page for documentation of the environment variables used by the GEOPM runtime that would otherwise be controlled by the wrapper script.

CPU Affinity Requirements
^^^^^^^^^^^^^^^^^^^^^^^^^

> The GEOPM runtime requires each of the application's MPI processes to be affinitized to different CPUs. This is a critical requirement for the runtime and must be enforced by the MPI launch command. When using the geopmlaunch wrapper, these affinity requirements are handled by geopmlaunch when the ``--geopm-affinity-enable`` command-line option is provided (see :doc:`geopmlaunch(1)<geopmlaunch.1>`). Otherwise, users must explicitly affinitize their application using the appropriate options for their chosen launcher.

Reviewer 2: ```suggestion
The GEOPM runtime requires each of the application's MPI processes to be affinitized to different CPUs. This is a critical requirement for the runtime and must be enforced by the MPI launch command. When using the ``geopmlaunch`` wrapper, these affinity requirements are handled by ``geopmlaunch`` when the ``--geopm-affinity-enable`` command-line option is provided (see :doc:`geopmlaunch(1)<geopmlaunch.1>`). Otherwise, users must explicitly affinitize their application using the appropriate options for their chosen launcher.
```

While the GEOPM control thread connects to the application it will automatically affinitize itself to the highest indexed core not used by the application if the application is not affinitized to a CPU on every core. If the application is using all cores of the system, the GEOPM control thread will be pinned to the highest logical CPU.

> Many ways exist to launch an MPI application, and no single uniform way of enforcing MPI rank CPU affinities can work across all job launch mechanisms. OpenMP runtimes, which are linked with compiler choice, also have different mechanisms for affinitizing OpenMP threads within CPUs available to each MPI process. The GEOPM control thread can also be launched as an application thread or process that can either be part of the primary MPI application or a completely different MPI application. Due to these factors, it is challenging to document the correct process affinitization across all configurations. Please refer to your site documentation about CPU affinity for the best solution for your system and consider extending the geopmlaunch wrapper to support your system configuration. For information on how to share these implementations with the community, refer to :doc:`contrib`.

Reviewer 2: ```suggestion
Many ways exist to launch an MPI application, and no single uniform way of enforcing MPI rank CPU affinities can work across all job launch mechanisms. OpenMP runtimes, which are linked with compiler choice, also have different mechanisms for affinitizing OpenMP threads within CPUs available to each MPI process. The GEOPM control thread can also be launched as an application thread or process that can either be part of the primary MPI application or a completely different MPI application. Due to these factors, it is challenging to document the correct process affinitization across all configurations. Please refer to your site documentation about CPU affinity for the best solution for your system and consider extending the ``geopmlaunch`` wrapper to support your system configuration. For information on how to share these implementations with the community, refer to :doc:`contrib`.
```

Resource Manager Integration
----------------------------

> The GEOPM Runtime package can seamlessly integrate with a compute cluster resource manager by altering the daemon of the resource manager running on the cluster compute nodes. An integration example with the SLURM resource manager through a SPANK plugin is available in the `geopm-slurm git repository <https://github.com/geopm/geopm-slurm>`_. This example aligns with the process described below.

Reviewer 2: Note: earlier in the documentation, the "Spank" capitalization was used, while here it's being capitalized as "SPANK". I would recommend using a consistent form throughout the documentation. Since it's an acronym, and seeing the Slurm documentation writes it as "SPANK", perhaps this form should be used.

For GEOPM's integration as the universal power management solution for a cluster, it's usual for a single agent algorithm with one policy to be applied across all compute nodes within a partition. The choice of agent rests upon the site's needs. For instance, if the aim is to keep the average CPU power draw for each node below a specific cap, the :doc:`power_balancer agent <geopm_agent_power_balancer.7>` is ideal. However, if the goal is to limit application CPU frequencies with exceptions for specific high-priority processes, the :doc:`frequency_map agent <geopm_agent_frequency_map.7>` is the best fit. Sites can also deploy a custom agent plugin. In every scenario, invoking ``geopm_agent_enforce_policy()`` before releasing compute resources ensures the enforcement of static limits impacting all user applications. For dynamic runtime features, users must initiate their MPI application using the :doc:`geopmlaunch(1) <geopmlaunch.1>` tool.

> To illustrate, if a system administrator wants to use the power_balancer agent, the process would involve setting a static power cap for apps not utilizing geopmlaunch, while optimizing power caps for performance when geopmlaunch is in use. The administrator would initiate the following JSON object in the compute node's boot image at ``/etc/geopm/environment-override.json``:

Reviewer 2: ```suggestion
To illustrate, if a system administrator wants to use the ``power_balancer`` agent, the process would involve setting a static power cap for apps not utilizing ``geopmlaunch``, while optimizing power caps for performance when ``geopmlaunch`` is in use. The administrator would install the following JSON object in the compute node's boot image at ``/etc/geopm/environment-override.json``:
```

.. code-block:: json

   {"GEOPM_AGENT": "power_balancer",
    "GEOPM_POLICY": "/shared_fs/config/geopm_power_balancer.json"}

> The controlling value, "CPU_POWER_LIMIT", is defined in a separate "geopm_power_balancer.json" file that could reside on a shared file system. This file can be generated using the :doc:`geopmagent(1) <geopmagent.1>` tool. By placing the policy file on a shared file system, you allow modifications to the limit without affecting the compute node boot image. Changing the policy value affects all new GEOPM processes but leaves running GEOPM processes untouched.

Reviewer 2: ```suggestion
The controlling value, ``CPU_POWER_LIMIT``, is defined in a separate "geopm_power_balancer.json" file that could reside on a shared file system. This file can be generated using the :doc:`geopmagent(1) <geopmagent.1>` tool. By placing the policy file on a shared file system, you allow modifications to the limit without affecting the compute node boot image. Changing the policy value affects all new GEOPM processes, but leaves running GEOPM processes untouched.
```

The GEOPM Runtime package installs the ``geopmlaunch`` command. This
command is a wrapper for MPI launch commands such as ``srun``, ``aprun``,
and ``mpiexec``, where the wrapper script enables the GEOPM runtime. The
``geopmlaunch`` command supports the same command-line interface as the
underlying launch command, while extending the interface with GEOPM-specific
options. The ``geopmlaunch`` application launches the primary compute
application and the GEOPM control thread on each compute node, and manages
all process CPU affinity requirements. This wrapper is documented in the
:doc:`geopmlaunch(1)<geopmlaunch.1>` man page.

``geopmlaunch`` supports various underlying MPI application launchers
as shown in the :doc:`geopmlaunch(1)<geopmlaunch.1>` man page. If your
system's launch mechanism is not supported, then you must enforce affinity
requirements, and all options to the GEOPM runtime must be passed through
environment variables. Please consult the :doc:`geopm(7)<geopm.7>` man page
for documentation of the environment variables used by the GEOPM runtime
that would otherwise be controlled by the wrapper script.

CPU Affinity Requirements
^^^^^^^^^^^^^^^^^^^^^^^^^

> The GEOPM runtime requires each of the application's MPI processes to

Reviewer 3: This whole section still needs to be reworked.  Too MPI centric.

automatically affinitize itself to the highest indexed core not used by the
application if the application is not affinitized to a CPU on every core. If
the application is using all cores of the system, the GEOPM control thread
will be pinned to the highest logical CPU.

Many ways exist to launch an MPI application, and no single uniform
way of enforcing MPI rank CPU affinities can work across all job launch
mechanisms. OpenMP runtimes, which are linked with compiler choice, also have
different mechanisms for affinitizing OpenMP threads within CPUs available
to each MPI process. The GEOPM control thread can also be launched as an
application thread or process that can either be part of the primary MPI
application or a completely different MPI application. Due to these factors,
it is challenging to document the correct process affinitization across all
configurations. Please refer to your site documentation about CPU affinity for
the best solution for your system and consider extending the ``geopmlaunch``
wrapper to support your system configuration. For information on how to
share these implementations with the community, refer to :doc:`contrib`.

> Resource Manager Integration

Reviewer 3: Remove this section unless we're going to test it.

all process CPU affinity requirements. This wrapper is documented in the
:doc:`geopmlaunch(1)<geopmlaunch.1>` man page.

> ``geopmlaunch`` supports various underlying MPI application launchers

Reviewer 3: The bla command
